src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java

```diff
@@ -0,0 +1,178 @@
+  public void handle(Message<JsonObject> message) {
+    JsonObject messageFromEventBus = message.body();
+
+    logger.info("Received message body: {}", messageFromEventBus);
+
+    Map<String, String> okapiHeaders = org.folio.orders.utils.HelperUtils.getOkapiHeaders(message);
+    HttpClientInterface httpClient = getHttpClient(okapiHeaders, true);
+
+    String pieceIdUpdate = messageFromEventBus.getString(PIECE_ID_UPDATE);
+
+    List<CompletableFuture<String>> futures = new ArrayList<>();
+    CompletableFuture<String> future = new VertxCompletableFuture<>(ctx);
+
+    futures.add(future);
+
+    // 1. Get updated piece by id from storage
+    getPieceById(pieceIdUpdate, lang, httpClient, ctx, okapiHeaders, logger).thenAccept(jsonPiece -> {
+      Piece piece = jsonPiece.mapTo(Piece.class);
+      ReceivingStatus receivingStatusAfterUpdate = piece.getReceivingStatus();
+      ReceivingStatus receivingStatusBeforeUpdate = ReceivingStatus.fromValue(messageFromEventBus.getString(RECEIVING_STATUS_BEFORE_UPDATE));
+
+      // 2. if receivingStatus is different - before writing to storage vs after writing to storage
+      if (receivingStatusBeforeUpdate.compareTo(receivingStatusAfterUpdate) != 0) {
+        String poLineId = piece.getPoLineId();
+        String query = String.format(PIECES_ENDPOINT, poLineId, LIMIT);
+
+        // 3. Get all pieces for poLineId above
+        getPieces(query, httpClient, okapiHeaders, logger).thenAccept(piecesCollection -> {
+          List<org.folio.rest.acq.model.Piece> listOfPieces = piecesCollection.getPieces();
+          
+          // 4. Get PoLine for the poLineId which will used to calculate PoLineReceiptStatus
+          getPoLineById(poLineId, lang, httpClient, ctx, okapiHeaders, logger).thenAccept(poLineJson -> {
+            PoLine poLine = poLineJson.mapTo(PoLine.class);
+            calculatePoLineReceiptStatus(poLine, listOfPieces)
+              .thenCompose(status -> updatePoLineReceiptStatus(poLine, status, httpClient, ctx, okapiHeaders, logger))
+              .thenAccept(future::complete);
+          })
+            .exceptionally(e -> {
+              logger.error("The error getting poLine by id {}", poLineId, e);
+              future.completeExceptionally(e);
+              return null;
+            });
+        })
+          .exceptionally(e -> {
+            logger.error("The error happened getting all pieces by poLine {}", poLineId, e);
+            future.completeExceptionally(e);
+            return null;
+          });
+      } else {
+        logger.debug("Receipt status of pieces and po line are consistent");
+        future.complete(null);
+      }
+    })
+      .exceptionally(e -> {
+        logger.error("The error happened getting a piece by id {}", pieceIdUpdate, e);
+        future.completeExceptionally(e);
+        return null;
+      });
+
+    // Now wait for all operations to be completed and send reply
+    completeAllFutures(ctx, httpClient, futures, message);
```

Comment:
In case the logic of `ReceiptStatusConsistency` updated POL's `receiptStatus` to `Fully Received` (or wise versa - from `Fully Received` to `Partially Received` or `Awaiting Receipt`), potentially entire order status can be changed as well. I believe these cases should be covered as well and an event to `org.folio.orders.order.update.status` should be sent...

Reply:
Should this event be sent from `AbstractHelper#updatePoLineReceiptStatus` where status changes to `FULLY_RECEIVED` or `AWAITING_RECEIPT` or `PARTIALLY_RECEIVED` ?

Reply:
As far as I can see current implementation of receiving/check-in flow is unchanged and POL's status is being calculated as part of that flow (initially I thought that this jira was going move that logic to new handler).
As far as I understand in most of the cases order is going to have mainly one line. But our implementation assumes that order might have up to 500 PO lines (the POL number sequence is up to 999). Potentially there might be some libraries which are going to work with such big orders. 
When the pieces are being received might belong to different lines and in case there are a lot lines involved e.g. user receives all the resources for 100 POL's, the logic might trigger up to 100 events and `OrderStatus` handler will start fetching a lot of data from storage but only the last one might end up updating order's status...

Reply:
> As far as I can see current implementation of receiving/check-in flow is unchanged and POL's status is being calculated as part of that flow (initially I thought that this jira was going move that logic to new handler).
I agree we should consolidate the calculation 

> When the pieces are being received might belong to different lines and in case there are a lot lines involved e.g. user receives all the resources for 100 POL's, the logic might trigger up to 100 events and OrderStatus handler will start fetching a lot of data from storage but only the last one might end up updating order's status...
Good point.  We need to be careful about where/when we fire the events from.

Reply:
I believe this part should be covered as well

Reply:
I see that the poLines with receivingStatus are still updated during "Receiving/Check-in" instead of sending event. Are these changes going to be added to this PR?

Reply:
No, there will a separate PR for that.